ext/nnstreamer/tensor_decoder/tensordec-boundingbox.h

```diff
@@ -0,0 +1,607 @@
+
+class BoxProperties
+{
+  public:
+  virtual ~BoxProperties () = default;
+
+  /* mandatory methods */
+  virtual int setOptionInternal (const char *param) = 0;
+  virtual int checkCompatible (const GstTensorsConfig *config) = 0;
+  virtual GArray *decode (const GstTensorsConfig *config, const GstTensorMemory *input) = 0;
+
+  void setInputWidth (guint width)
+  {
+    i_width = width;
+  }
+  void setInputHeight (guint height)
+  {
+    i_height = height;
+  }
+  void setTotalLabels (guint labels)
+  {
+    total_labels = labels;
+  }
+
+  guint getInputWidth ()
+  {
+    return i_width;
+  }
+  guint getInputHeight ()
+  {
+    return i_height;
+  }
+
+  protected:
+  guint i_width; /**< Input Video Width */
+  guint i_height; /**< Input Video Height */
+
+  guint max_detection;
+  guint total_labels;
+};
+
+
+class BoundingBox
+{
+  public:
+  BoundingBox ();
+  ~BoundingBox ();
+
+  gboolean checkLabelProps ();
+  int setBoxDecodingMode (const char *param);
+  int setLabelPath (const char *param);
+  int setVideoSize (const char *param);
+  int setInputModelSize (const char *param);
+  void draw (GstMapInfo *out_info, GArray *results);
+  void logBoxes (GArray *results);
+  void updateCentroids (GArray *boxes);
+
+  int setOption (int opNum, const char *param);
+  GstCaps *getOutCaps (const GstTensorsConfig *config);
+  GstFlowReturn decode (const GstTensorsConfig *config,
+      const GstTensorMemory *input, GstBuffer *outbuf);
+
+  private:
+  bounding_box_modes mode;
+  BoxProperties *bdata;
+
+  /* From option2 */
+  imglabel_t labeldata;
+  char *label_path;
+
+  /* From option4 */
+  guint width; /**< Output Video Width */
+  guint height; /**< Output Video Height */
+
+  /* From option6 (track or not) */
+  gint is_track;
+  guint centroids_last_id; /**< The last_id of centroid valid id is 1, 2, ... (not 0). */
+  guint max_centroids_num; /**< The maximum number of centroids */
+  guint consecutive_disappear_threshold; /**< The threshold of consecutive disappeared frames */
+
+  GArray *centroids; /**< Array for centroids */
+  GArray *distanceArray; /**< Array for distances */
+
+  /* From option7 (log or not) */
+  gint do_log;
+
+  gboolean flag_use_label;
+};
+
+/** @brief Mathematic inverse of sigmoid function, aka logit */
+static float
+logit (float x)
+{
+  if (x <= 0.0f)
+    return -INFINITY;
+
+  if (x >= 1.0f)
+    return INFINITY;
+
+  return log (x / (1.0 - x));
+}
+
+#define MOBILENET_SSD_PARAMS_THRESHOLD_IDX 0
+#define MOBILENET_SSD_PARAMS_Y_SCALE_IDX 1
+#define MOBILENET_SSD_PARAMS_X_SCALE_IDX 2
+#define MOBILENET_SSD_PARAMS_H_SCALE_IDX 3
+#define MOBILENET_SSD_PARAMS_W_SCALE_IDX 4
+#define MOBILENET_SSD_PARAMS_IOU_THRESHOLD_IDX 5
+#define MOBILENET_SSD_PARAMS_MAX 6
+
+#define DETECTION_THRESHOLD (.5f)
+#define THRESHOLD_IOU (.5f)
+#define Y_SCALE (10.0f)
+#define X_SCALE (10.0f)
+#define H_SCALE (5.0f)
+#define W_SCALE (5.0f)
+
+class MobilenetSSD : public BoxProperties
```

Comment:
You do not need to declare subclasses in the header.
As long as we now can have separated subclasses, I recommend making separated files for subclasses and remove all info about each subclass in the header. (e.g., /ext/nnstreamer/tensor_decoder/boxproperties/mobilenetssd.cc)

Then, you can apply factory method in this base class that creates sub classes according to the property name. By properly designing this, you will not need the enums for each "subclass", too. You will need to think about how to accept "new" subclasses from the base class.

Reply:
The principle is to put commonly referenced classes and definitions in header.

If a class, #define, or a function is to be referenced by a single entity (e.g., by mobile-SSD only), you don't need to put it in a header.

Reply:
Moved subclasses to boxproperties subdir.
Thank you :)